src/app/components/shared/date-time-filter/filter.component.ts

```diff
@@ -0,0 +1,199 @@
+  dateFinishedBefore?: NgbDate;
+  todFiltering?: boolean;
+  todIgnoreDst?: boolean;
+  todStartedAfter?: string;
+  todFinishedBefore?: string;
+}
+
+@Component({
+  selector: "baw-date-time-filter",
+  templateUrl: "filter.component.html",
+  styleUrls: ["filter.component.scss"]
+})
+export class DateTimeFilterComponent extends withUnsubscribe() implements AfterViewInit {
+  public constructor(
+    private formatter: NgbDateParserFormatter,
+    private calendar: NgbCalendar
+  ) { super(); }
+
+  @ViewChild(NgForm) public form: NgForm;
+
+  @Input() public project: Project;
+  @Input() public region: Region;
+  @Input() public site: Site;
+  @Output() public constructedFilters = new EventEmitter<Filters>();
+
+  private previousFilters: string;
+  public model: Model = { todIgnoreDst: true };
+  public errors: {
+    todBoundaryError?: boolean;
+  } = {};
+
+  public ngAfterViewInit(): void {
+    this.form.valueChanges
+      .pipe(
+        debounceTime(defaultDebounceTime),
+        distinctUntilChanged(),
+        takeUntil(this.unsubscribe)
+      )
+      .subscribe((model: Model): void => this.updateHref(model));
+  }
+
+  public updateHref(model: Model): void {
+    // validate that the dates are a valid format & are within bounds before sending an API filter request
+    // has the value of null if there is no update to the filters
+    const newFilters = this.generateFilters(model);
+
+    // if the new filters will be the same as the old filters, do not emit a filter update event
+    if (newFilters && this.model) {
+      this.constructedFilters.emit(newFilters);
+    }
+  }
+
+  public isInvalidDate(date: string | NgbDate): boolean {
+    if (!date) {
+      return false;
+    } else if (typeof date === "string") {
+      return true;
+    }
+
+    const isValidDate = this.calendar.isValid(date)
+    return date?.year.toString().length < 4 || !isValidDate;
+  }
+
+  /**
+   * Validates that the start date comes before the end date
+   *
+   * @param formStartDate A date object of type `NgbDate` supplied by the bootstrap date input representing the start date
+   * @param formEndDate A date object of type `NgbDate` supplied by the bootstrap date input representing the end date
+   * @returns Whether the start date comes before the end date, or the two dates are equal
+   */
+  public dateIsInBoundary(formStartDate: NgbDate, formEndDate: NgbDate): boolean {
+    // if the user has not set a start date and end date, we know that the date range is not defined and therefore, cannot be out of bounds
+    if (!isInstantiated(formStartDate) || !isInstantiated(formEndDate)) {
+      return true;
+    }
+
+    // converting to a date object allows the comparison of dates using boolean operators
+    const startDate = new Date(formStartDate.year, formStartDate.month, formStartDate.day);
+    const endDate = new Date(formEndDate.year, formEndDate.month, formEndDate.day);
+
+    return startDate <= endDate;
+  }
+
```

Comment:
This is a good point, I'll update the code accordingly

Reply:
I've had a play around with angular form validations for these and I can transfer the `dateIsInBoundary` method to ng form validation, but not the isInvalid date.

This is because of some trickery to do with how the bootstrap directives work.

When it does not match bootstrap form validation (`[0-9]+-\d{2}-\d{2}`), it returns a string, however, when it matches the RE above, it returns an `NgbDate` object. This means that using ngForm validation attributes such as `pattern="\d{4}-\d{2}-\d{2}"` does not work (because it's trying to compare to an object), and `minlength="10"` will return true even when the date is < 10 characters because the object will be > 10.